feat: support  headline's cookie from TODOs

thuyen opened the pull request:
## Summary

Add function `update_todo_cookie` to update a headline's cookie from a TODO state change.

## Related Issues

<!-- Link issues that are related to this PR. You may link issues you think should be closed by this PR. -->

Closes #305


## Changes

<!-- List the major changes made in this PR. -->
![Screenshot from 2025-03-15 21-25-54](https://github.com/user-attachments/assets/ea45a48a-7e72-45e1-8e80-6d9c54f704e4)




## Checklist

I confirm that I have:

- [x] **Followed the
      [Conventional Commits](https://www.conventionalcommits.org/)
      specification** (e.g., `feat: add new feature`, `fix: correct bug`,
      `docs: update documentation`).
- [x] **My PR title also follows the conventional commits specification.**
- [x] **Updated relevant documentation,** if necessary.
- [x] **Thoroughly tested my changes.**
- [x] **Added tests** (if applicable) and verified existing tests pass with
      `make test`.
- [x] **Checked for breaking changes** and documented them, if any.

kristijanhusak commented:
Lets rebase and then I'll take a look. Thanks!

kristijanhusak commented:
Generally looks good, and works as expected as long as only sub-todos are used.
The only issue is when both checkboxes and sub-todos are used. For example, having this content:

 ```
* Test [/]
- [ ] foo
- [ ] bar
** TODO Test 1
** TODO Test 2
** TODO Test 3
```

Toggling either sub-todo or a checkbox, ends up adding `[1/5]` to the top level headline.
In emacs, it seems that checkbox list has a priority over sub-todos.
For example, given the content above, if you change the sub-todo to DONE, it will update cookie to `[1/3]`, but if you recalculate the cookie value with `C-c #`, it will actually take checkboxes state and update the cookie back to `[0/2]`.
If you end up having one checkbox checked and cookie value of `[1/2]`, switching a sub-todo to done will again update cookie to `[1/3]`. I'm assuming Emacs orgmode is not handling this conflict since it's highly unlikely that anyone is using it like this.
Nontheless, I'd like to follow the same logic. If there are checkboxes in the headline body, prioritize that over sub-tasks.

thuyen commented:
> Generally looks good, and works as expected as long as only sub-todos are used. The only issue is when both checkboxes and sub-todos are used. For example, having this content:
> 
> ```
> * Test [/]
> - [ ] foo
> - [ ] bar
> ** TODO Test 1
> ** TODO Test 2
> ** TODO Test 3
> ```
> 
> Toggling either sub-todo or a checkbox, ends up adding `[1/5]` to the top level headline. In emacs, it seems that checkbox list has a priority over sub-todos. For example, given the content above, if you change the sub-todo to DONE, it will update cookie to `[1/3]`, but if you recalculate the cookie value with `C-c #`, it will actually take checkboxes state and update the cookie back to `[0/2]`. If you end up having one checkbox checked and cookie value of `[1/2]`, switching a sub-todo to done will again update cookie to `[1/3]`. I'm assuming Emacs orgmode is not handling this conflict since it's highly unlikely that anyone is using it like this. Nontheless, I'd like to follow the same logic. If there are checkboxes in the headline body, prioritize that over sub-tasks.

I reverted `update_cookie` to just handle checked boxes; and added `update_todo_cookie` to handle TODO state change separately. At least for the example above, it works as you would expect.

kristijanhusak reviewed (approved):
Looks good, thanks!
